packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,396 @@
+        }
+      }
+    }
+  }
+}
+
+/**
+ * From the given layout types, gets the struct member matching the given struct canonical name and struct member label.
+ */
+function getStructMemberFromLayoutTypes(
+  types: Record<string, TypeItem<string>>,
+  structCanonicalName: string,
+  memberLabel: string,
+) {
+  const structType = findTypeWithLabel(types, `struct ${structCanonicalName}`);
+  const structMembers = structType?.members;
+  if (structMembers !== undefined) {
+    for (const structMember of structMembers) {
+      assert(typeof structMember !== 'string');
+      if (structMember.label === memberLabel) {
+        return structMember;
+      }
+    }
+  }
+  return undefined;
+}
+
+/**
+ * From the given layout types, gets the type matching the given type label.
+ */
+function findTypeWithLabel(types: Record<string, TypeItem>, label: string) {
+  for (const type of Object.values(types)) {
+    if (type.label === label) {
+      return type;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Makes a modified copy of the solc input to add state variables in each contract for namespaced struct definitions,
+ * so that the compiler will generate their types in the storage layout.
+ *
+ * This deletes all functions for efficiency, since they are not needed for storage layout.
+ * We also need to delete modifiers and immutable variables to avoid compilation errors due to deleted
+ * functions and constructors.
+ *
+ * Also sets the outputSelection to only include storageLayout and ast, since the other outputs are not needed.
+ *
+ * @param input The original solc input.
+ * @param output The original solc output.
+ * @returns The modified solc input with storage layout that includes namespaced type information.
+ */
+export function makeNamespacedInputCopy(input: SolcInput, output: SolcOutput): SolcInput {
+  const modifiedInput: SolcInput = JSON.parse(JSON.stringify(input));
+
+  modifiedInput.settings = {
+    outputSelection: {
+      '*': {
+        '*': ['storageLayout'],
+        '': ['ast'],
+      },
+    },
+  };
+
+  for (const [sourcePath] of Object.entries(modifiedInput.sources)) {
+    const source = modifiedInput.sources[sourcePath];
+    if (source.content === undefined) {
+      continue;
+    }
+
+    // Collect all contract definitions
+    const contractDefs = [];
+    for (const contractDef of findAll('ContractDefinition', output.sources[sourcePath].ast)) {
+      contractDefs.push(contractDef);
+    }
+
+    // Iterate backwards so we can delete source code without affecting remaining indices
+    for (let i = contractDefs.length - 1; i >= 0; i--) {
+      const contractDef = contractDefs[i];
+      const nodes = contractDef.nodes;
+      for (let j = nodes.length - 1; j >= 0; j--) {
+        const node = nodes[j];
+        if (
+          isNodeType('FunctionDefinition', node) ||
+          isNodeType('ModifierDefinition', node) ||
+          (isNodeType('VariableDeclaration', node) && node.mutability === 'immutable')
```

Comment:
I'm realizing this is more complex than I said initially. You probably ran into this. If we delete functions, we can no longer have references to those functions, so we have to delete every possible place where a reference would show up, i.e. every possible expression. I would delete all `VariableDeclaration` nodes (ignoring the mutability), because they can contain an initialization expression. The other thing that isn't covered by your code here is initialization in the parents list:

```solidity
contract A is B(expr()) {
}
```

This is in `contractDef.baseContracts[i].arguments`, but this is an array and we don't have bounds that include parenthesis. What I would do is take each one of `contractDef.baseContracts[i]` and replace all text within the node bounds with the value of `contractDef.baseContracts[i].baseName`.